### PR TITLE
[SHELL32_APITESTS] Add ROS.ico test file. Improve shell32:ExtractIconEx tests

### PR DESCRIPTION
**Pull request by Doug-Lyons**
## Purpose

_Improve shell32:ExtractIconEx tests by updating it and adding more test files and 1 new icon file._

JIRA issue: [ROSTESTS-381](https://jira.reactos.org/browse/ROSTESTS-381)

## Proposed changes

_Add new ROS.ico file which has both a normal icon and a PNG one._
_Update testing to test for count of icons in file and separately extract the 0th one._
_Add tests for sysicon.ico, explorer.exe and the new ROS.ico files._

WinWP:
WinXP_ExtractIconEx

W2K3SP2:
W2K3SP2_ExtractIconEx

### CODE REVIEW

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp [hunk 1]
```diff
@@ -22,16 +22,33 @@ EXTRACTICONTESTS IconTests[] =
...
+    {L"%SystemRoot%\\non-existent-file.sdf", 0},
+
+    /* Multiple icons in the same ICO file (6 icons) */
+    {L"%SystemRoot%\\bin\\sysicon.ico", 1},
+
+    /* Multiple icons in the same EXE file (18 icons) */
+    {L"%SystemRoot%\\explorer.exe", 18},
+
+    /* One standard icon and one PNG in same file */
+    {L"%SystemRoot%\\bin\\ROS.ico", 1}
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ExtractIconEx.cpp:34]: Can you embed `ROS.ico` and `sysicon.ico` into the resource?
Are the `explorer.exe` tests compatible to WinXP/Win2k3?

Comment (katahiromz) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp:34]: https://github.com/reactos/reactos/blob/9686771fc07fd9bb9c77344f88c5fc1f863f5fe3/modules/rostests/apitests/shell32/resource.rc#L7

Comment (katahiromz) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp:34]:
```txt
2 RCDATA "ROS.ico"
3 RCDATA "sysicon.ico"
```

Comment (katahiromz) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp:34]:
```c 
HRSRC hRes = FindResourceW(NULL, MAKEINTRESOURCEW(2), RT_RCDATA);
HGLOBAL hData = LoadResource(NULL, hRes);
...
```

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp [hunk 2]
```diff
@@ -22,16 +22,33 @@ EXTRACTICONTESTS IconTests[] =
...
 
     /* Non-existing files */
-    {L"%SystemRoot%\\non-existent-file.sdf", 0}
+    {L"%SystemRoot%\\non-existent-file.sdf", 0},
+
+    /* Multiple icons in the same ICO file (6 icons) */
+    {L"%SystemRoot%\\bin\\sysicon.ico", 1},
+
+    /* Multiple icons in the same EXE file (18 icons) */
+    {L"%SystemRoot%\\explorer.exe", 18},
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ExtractIconEx.cpp:31]: Are the explorer.exe tests compatible to WinXP/Win2k3?

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp (hunk 1 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp:34]: @katahiromz, I have looked into how to do this and I am not able to figure it out for now. Maybe you can offer a suggestion or maybe just leave it for the next PR? Thanks.

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp (hunk 2 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp:31]: @katahiromz, Yes, they are compatible with WinXP/Win2k3. I have attached screen shots in the Description above to demonstrate. I did have to add the two ROS.ico and sysicon.ico to the C:\Windows\bin directory first though.

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp (hunk 1 above)
Comment (HBelusca) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp:34]: > ```c
> HRSRC hRes = FindResourceW(NULL, MAKEINTRESOURCEW(2), RT_RCDATA);
> HGLOBAL hData = LoadResource(NULL, hRes);
> ...
> fclose(fout);
> ```

Huh just use LoadIcon or LoadImage ...

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp (hunk 2 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp:31]: Note: Explorer.exe in WinXP/Win2k3 has 18 embedded icons just like the one in ReactOS.

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp (hunk 1 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp]: OK, @katahiromz and @HBelusca, I believe that I now have the icons embedded in the EXE file as was suggested. Please review again and offer your comments. Thanks.

**Comment by Doug-Lyons** [comment]
For some reason, github will not allow me to request reviews from both @katahiromz and @HBelusca  without deleting the review request from the other. Unless there is something that I am doing wrong. Sorry for the confusion. I am open to suggestions. Thanks.

**Review by Doug-Lyons — COMMENTED** [review]
I have some "pending" items and hopefully this will fix them.

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp [hunk 3]
```diff
@@ -22,16 +24,72 @@ EXTRACTICONTESTS IconTests[] =
...
-    UINT i, nReturnedIcons;
+    UINT i, nReturnedIcons, nExtractedIcons, iSize;
+    HRSRC hRes;
+    HGLOBAL hData;
+    LPVOID lpResLock;
+    LPCSTR sData;
+    FILE *fout;
+    CHAR FileName[2][13] = { "ROS.ico", "sysicon.ico" };
+
+    for (i = 2; i < 4; i++) // Point to resources 2 and 3
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ExtractIconEx.cpp:50]: Could you split this task (extracting embeded data) into newly defined functions?

Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp]: @katahiromz, I have split out the extraction of embedded data and added deleting of the files as you asked. Thanks.

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp [hunk 4]
```diff
@@ -6,13 +6,62 @@
...
+    // Lock the resource into global memory.
+    lpResLock = LockResource(hData);
+    if (lpResLock == NULL)
+    {
+        skip("Could not lock resource (%d).", i);
+        return FALSE;
+    }
+
+    // Get location of resource in memory
+    sData = (LPSTR)lpResLock;
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ExtractIconEx.cpp:58]:
```suggestion
```

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp [hunk 5]
```diff
@@ -6,13 +6,62 @@
...
+
+    fout = fopen(FileName,"wb");
+    fwrite(sData, iSize, 1, fout);
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ExtractIconEx.cpp:61]:
```suggestion
    fwrite(lpResLock, iSize, 1, fout);
```

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp [hunk 6]
```diff
@@ -22,16 +71,44 @@ EXTRACTICONTESTS IconTests[] =
...
     }
+
+    /* Check if the 0th icon can be extracted successfully */
+    for (i = 0; i < _countof(IconTests); ++i)
+    {
+        nExtractedIcons = ExtractIconExW(IconTests[i].pszFilePath, 0, NULL, NULL, 1);
+        ok(nExtractedIcons == IconTests[i].nIcons, "ExtractIconExW(%u): Expects %u icons, got %u\n", i, IconTests[i].nIcons, nExtractedIcons);
+    }
+
+    remove(FileName[0]);
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ExtractIconEx.cpp:111]: Use `DeleteFileA`.

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp [hunk 7]
```diff
@@ -6,13 +6,62 @@
...
+    if (lpResLock == NULL)
+    {
+        skip("Could not lock resource (%d).", i);
+        return FALSE;
+    }
+
+    // Get location of resource in memory
+    sData = (LPSTR)lpResLock;
+
+    fout = fopen(FileName,"wb");
```
[the remark this answers is not in the document]
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp]: Done. Thanks.

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp [hunk 8]
```diff
@@ -6,13 +6,62 @@
...
     PCWSTR pszFilePath;
     UINT nIcons;
 } EXTRACTICONTESTS;
 
+BOOL ResourceToFile(INT i, CHAR *FileName)
+{
+    FILE *fout;
+    HGLOBAL hData;
+    HRSRC hRes;
+    LPCSTR sData;
```
[the remark this answers is not in the document]
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp]: Done. Thanks.

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp (hunk 4 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp]: Done. Thanks.

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp (hunk 5 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp]: Done. Thanks.

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp (hunk 6 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp]: Done. Thanks.

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp [hunk 9]
```diff
@@ -3,16 +3,64 @@
...
 
+BOOL ResourceToFile(INT i, CHAR *FileName)
+{
+    FILE *fout;
+    HGLOBAL hData;
+    HRSRC hRes;
+    LPVOID lpResLock;
+    UINT iSize;
+
+//    trace("i is '%d' and Filename is '%s'.\n", i, FileName);
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ExtractIconEx.cpp:27]:
```suggestion
```

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp [hunk 10]
```diff
@@ -3,16 +3,64 @@
...
+#include <stdio.h>
+#include <io.h>
 
 typedef struct
 {
     PCWSTR pszFilePath;
     UINT nIcons;
 } EXTRACTICONTESTS;
 
+BOOL ResourceToFile(INT i, CHAR *FileName)
```
Comment (katahiromz) [inline modules/rostests/apitests/shell32/ExtractIconEx.cpp:19]:
```suggestion
BOOL ResourceToFile(INT i, LPCSTR FileName)
```

**Review by katahiromz — APPROVED** [review]

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp (hunk 9 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp]: Done. Thanks.

File: modules/rostests/apitests/shell32/ExtractIconEx.cpp (hunk 10 above)
Comment (Doug-Lyons) [reply modules/rostests/apitests/shell32/ExtractIconEx.cpp]: Done. Thanks.

**Review by katahiromz — APPROVED** [review]
It's go.

**Comment by JoachimHenze** [comment]
Ftr, not mentioned in the description but there were even failing tests **before** this PR in that test:
https://reactos.org/testman/compare.php?ids=86524 VBox
https://reactos.org/testman/compare.php?ids=86525 KVM

**Comment by JoachimHenze** [comment]
There is one thing that I do not understand yet, and that I want to understand before accepting the PR:
**Before** your changes there were 3 tests.
WHS passed 3/3 https://reactos.org/testman/compare.php?ids=86181
VBox only 2/3 https://reactos.org/testman/compare.php?ids=86524
KVM only 2/3 https://reactos.org/testman/compare.php?ids=86525

After your changes all: Ros and WHS does succeed all tests. So it appears to me at first glance that the tests are more relaxed now, as in "hiding differences between ros and WHS that are actually there".
But I didn't spot in your code changes yet, where you (weakened or fixed) the already existing test!
Can you elaborate or explain that?

**Comment by Doug-Lyons** [comment]
Testbot results of current PR:
ExtractIconEx_improve_07u.patch JID65462 on top of 0.4.15-dev-5858-g16decc6

VBox: https://reactos.org/testman/compare.php?ids=86528,86532
KVM:  https://reactos.org/testman/compare.php?ids=86527,86531

The single failure is still there and in the new results, it is the first line as follows:
ExtractIconEx.cpp#l106: Test failed: ExtractIconExW(0): Expects 1 icons, got 0

**Review by JoachimHenze — APPROVED** [review]

**Review by ericsomto — APPROVED** [review]

**Pull request merged**